File: tests/lua/xpb_security_test.lua
-- XPB V2 Lua Runtime Security Validation (post-hardening)
--
-- Each test exercises a class of input that pre-hardening either succeeded
-- silently (returning garbage) or hung the interpreter. After the runtime
-- gained ensure_bytes() + read_array_count(element_min_bytes, max_elements),
-- every one of these inputs now raises a clear Lua error. The script
-- asserts that error is raised — if a future change removes the guard, a
-- test goes from "errored as expected" back to "succeeded silently" and
-- the script fails.

local xpb = require("xpb")

local passed = 0
local failed = 0

local function expect_error(name, fn)
    local ok, err = pcall(fn)
    if not ok then
        print(string.format("  [PASS] %s (errored: %s)", name, err))
        passed = passed + 1
    else
        print(string.format("  [FAIL] %s — call succeeded; expected an error", name))
        failed = failed + 1
    end
end

print("===========================================")
print("XPB V2 Lua Runtime Security Validation")
print("===========================================")

-- XPB-109: read_bool over a 0-byte buffer must raise, not return true.
print("\n=== XPB-109: read_bool past EOF must error ===")
expect_error("read_bool over empty buffer", function()
    local dec = xpb.Decoder("")
    return dec:read_bool()
end)

-- XPB-109: read_int32 over a too-short buffer must raise.
print("\n=== XPB-109: read_int32 past EOF must error ===")
expect_error("read_int32 over 3-byte buffer", function()
    local dec = xpb.Decoder("\x01\x02\x03")
    return dec:read_int32()
end)

-- XPB-108: read_array_int32 requires an explicit max — calling it without
-- the new argument should error.
print("\n=== XPB-108: read_array_int32 requires explicit max ===")
expect_error("read_array_int32 with no max arg", function()
    local enc = xpb.Encoder(8)
    enc:write_int32(1000)
    local dec = xpb.Decoder(enc:finish())
    return dec:read_array_int32()
end)

-- XPB-108: even with a max, an oversized wire count is rejected.
print("\n=== XPB-108: oversized count rejected upfront ===")
expect_error("read_array_int32 with count > max", function()
    local enc = xpb.Encoder(8)
    enc:write_int32(1000) -- 1000 elements claimed
    local dec = xpb.Decoder(enc:finish())
    return dec:read_array_int32(64) -- caller's budget is 64
end)

-- XPB-108: buffer-bound check still works (4-byte buffer can't hold 1000
-- int32 elements).
print("\n=== XPB-108: count exceeding buffer is rejected ===")
expect_error("read_array_int32 with count > buffer", function()
    local enc = xpb.Encoder(8)
    enc:write_int32(1000)
    local dec = xpb.Decoder(enc:finish())
    return dec:read_array_int32(1 << 20) -- huge max, but buffer is too small
end)

-- XPB-109: skip past EOF must error.
print("\n=== XPB-109: skip(n) past EOF must error ===")
expect_error("skip past buffer end", function()
    local dec = xpb.Decoder("ab")
    dec:skip(100000)
end)

-- Sanity: legitimate reads still work.
print("\n=== Regression: legitimate round-trip still works ===")
local function legit_roundtrip()
    local enc = xpb.Encoder(64)
    enc:write_int32(3)
    enc:write_int32(10)
    enc:write_int32(20)
    enc:write_int32(30)
    local dec = xpb.Decoder(enc:finish())
    local arr = dec:read_array_int32(16) -- caller budget = 16
    return arr[1] == 10 and arr[2] == 20 and arr[3] == 30
end
if legit_roundtrip() then
    print("  [PASS] honest count + elements decoded correctly")
    passed = passed + 1
else
    print("  [FAIL] legitimate payload no longer round-trips")
    failed = failed + 1
end

print(string.format("\nResults: %d passed, %d failed", passed, failed))
os.exit(failed > 0 and 1 or 0)

File: tests/java/XpbSecurityTest.java
package xpb;

import java.io.ByteArrayOutputStream;

/**
 * XPB V2 Java Runtime Security Validation (post-hardening).
 *
 * Each test exercises a class of input that pre-hardening either succeeded
 * silently or hit an uncatchable JVM allocation. After the runtime gained
 * readArrayCount(elementMinBytes, maxElements), every one of these inputs
 * now raises a clear RuntimeException. The tests assert that the
 * exception fires — a future change that removes the guard takes a test
 * from "rejected" back to "silently allocated GB of int[]" and fails.
 */
public class XpbSecurityTest {
    private static int passed = 0;
    private static int failed = 0;

    private static void expectThrow(String name, Runnable fn) {
        try {
            fn.run();
            System.out.println("  [FAIL] " + name + " — call succeeded; expected an exception");
            failed++;
        } catch (RuntimeException e) {
            System.out.println("  [PASS] " + name + " — threw: " + e.getMessage());
            passed++;
        }
    }

    private static byte[] int32LE(int v) {
        byte[] out = new byte[4];
        out[0] = (byte) (v & 0xFF);
        out[1] = (byte) ((v >> 8) & 0xFF);
        out[2] = (byte) ((v >> 16) & 0xFF);
        out[3] = (byte) ((v >> 24) & 0xFF);
        return out;
    }

    public static void main(String[] args) {
        System.out.println("===========================================");
        System.out.println("XPB V2 Java Runtime Security Validation");
        System.out.println("===========================================");

        // XPB-107: readArrayInt32 requires explicit maxElements. The old
        // signature took no max — the new one takes one and rejects an
        // oversized wire-supplied count up front.
        System.out.println("\n=== XPB-107: oversized count rejected against caller's max ===");
        expectThrow("readArrayInt32 with count > max", () -> {
            byte[] payload = int32LE(1000); // claims 1000 elements
            Decoder dec = new Decoder(payload);
            dec.readArrayInt32(64); // caller's budget is 64
        });

        // XPB-107: negative count rejected.
        System.out.println("\n=== XPB-107: negative count rejected ===");
        expectThrow("readArrayInt32 with negative count", () -> {
            byte[] payload = int32LE(-1);
            Decoder dec = new Decoder(payload);
            dec.readArrayInt32(1024);
        });

        // XPB-107: even within the max, a count that can't fit in the
        // remaining buffer is rejected (buffer-bound check still fires).
        System.out.println("\n=== XPB-107: count exceeding buffer is rejected ===");
        expectThrow("readArrayInt32 with count > remaining buffer", () -> {
            byte[] payload = int32LE(1000);
            Decoder dec = new Decoder(payload);
            dec.readArrayInt32(1 << 20); // huge max, but buffer is 4 bytes
        });

        // XPB-107: max < 0 is itself an error (caller bug).
        System.out.println("\n=== XPB-107: negative max rejected with IllegalArgumentException ===");
        try {
            byte[] payload = int32LE(0);
            Decoder dec = new Decoder(payload);
            dec.readArrayInt32(-5);
            System.out.println("  [FAIL] negative max accepted");
            failed++;
        } catch (IllegalArgumentException e) {
            System.out.println("  [PASS] negative max rejected: " + e.getMessage());
            passed++;
        }

        // Sanity: legitimate decode round-trips under an honest max.
        System.out.println("\n=== Regression: legitimate array round-trip works ===");
        try {
            Encoder enc = new Encoder(64);
            int[] in = {10, 20, 30, 40};
            enc.writeInt32(in.length);
            for (int v : in) enc.writeInt32(v);
            byte[] data = enc.finish();

            Decoder dec = new Decoder(data);
            int[] out = dec.readArrayInt32(16); // caller's max
            if (out.length == 4 && out[0] == 10 && out[3] == 40) {
                System.out.println("  [PASS] round-trip preserved");
                passed++;
            } else {
                System.out.println("  [FAIL] round-trip corrupted");
                failed++;
            }
        } catch (Exception e) {
            System.out.println("  [FAIL] round-trip threw: " + e.getMessage());
            failed++;
        }

        System.out.println("\nResults: " + passed + " passed, " + failed + " failed");
        if (failed > 0) System.exit(1);
    }
}
